Add lbipam support for shared ips

usiegl00 opened the pull request:
This disables the checks that prevent the sharing of ips between services.

Fixes: #21776

dylandreimerink commented on operator/pkg/lbipam/service_store.go:
```diff
@@ -77,6 +83,28 @@ type ServiceView struct {
+			}
+		}
+	}
+	// They both use the Cluster external traffic policy, or they both point to the exact same set of pods (i.e. the pod selectors are identical).
```
> or they both point to the exact same set of pods (i.e. the pod selectors are identical).

There is a little bit more to this check than currently implemented. There are a few missing things:
* Services are namespaced, so you also have to make sure the services are in the same namespace in addition to the selectors.
* Services can have `selectors` but there are also other ways to associate endpoints/pods with services, namely with `Endpoints`, `EndpointSlices` and their Cilium specific variants. 

Implementing support for the separate endpoint objects is hard, as in a lot of work to get it right, so I suggest you to not support those.

So enforce that service with sharing-keys have to have a `Selector` and document the limitation, preferably even add a github issue for yourself or someone else to add this in a followup PR.

dylandreimerink commented on operator/pkg/lbipam/service_store.go:
```diff
@@ -77,6 +83,28 @@ type ServiceView struct {
+				return false
+			}
+		}
+	}
```
`.Ports` is of type `[]slim_core_v1.ServicePort`, so `port1` and `port2` are not port numbers, they are structures including `.Name`, `.Protocol`, `.Port`, and `.NodePort`. The current check would allow overlapping ports as long as the name is different or overlapping ports if they have different protocols which we currently don't support. 

There is also the matter of `.NodePort`. Its description says that it can be used for LoadBalancer services:
> The port on each node on which this service is exposed when type is NodePort or LoadBalancer.

But I am not sure if we need to check it here, given that node ports globally can't overlap (so I would expect the API server to have already rejected a sevice with the same nodeport as another).
```suggestion
	for _, port1 := range sv.Ports {
		for _, port2 := range osv.Ports {
			if port1.Port == port2.Port {
				return false
			}
		}
	}
```

dylandreimerink commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -604,22 +615,25 @@ func (ipam *LBIPAM) stripOrImportIngresses(sv *ServiceView) (statusModified bool
 
-			err = lbRange.alloc.Alloc(ip, true)
+			serviceViews := make([]*ServiceView, 0, 1)
+			serviceViews = append(serviceViews, sv)
```
Small nit, but you can write this faster

```suggestion
			serviceViews := []*ServiceView{sv}
```

dylandreimerink commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -710,6 +755,44 @@ func (ipam *LBIPAM) satisfyServices(ctx context.Context) error {
+				break
+			}
+		}
+	}
```
Not sure this is right. We are not taking the requests of the service into account. This logic finds the first IP within the IP sharing group and assigns it. So if the group only has an IPv6 and the service requests only an IPv4 IP it will get the IPv6 address anyway.

Or if a service requests a very specific IPv4 address, it will get whatever IP from the sharing group and then potentially also the IP it actually requested.

I think the sharing key logic should be more integrated with the logic to give the service what it wants which is:
* A list of specific IPs
* an arbitrary IPv4
* an arbitrary IPv6
* an arbitrary IPv4 and an arbitrary IPv4

These requests are not in addition to the sharing key, they should be combined. 

In the case of a list of specific IPs an additional check must be added that the existing service with that IP and the new one has the same key and is compatible.

In the case of the non specific requests we should first attempt to satisfy the requests from the shared group, and if not possible fallback to the normal process.

dylandreimerink reviewed (changes requested):
Thank you so much for the work you have put into this PR! Overall really nice, though I do think I spotted some things that need to be addressed

dylandreimerink reviewed (changes requested):
@usiegl00 thanks again for making the adjustments. I have been playing these changes a bit to test things out and found some bugs. While debugging I concluded that the code had become hard to read due to the functions becoming to big so I took it upon myself to do a bit of restructuring. I figured that would be easier than to communicate all of my nit picks. I forked your branch and added 3 commits atop with tests for the bug, the bugfix and my structural changes https://github.com/dylandreimerink/cilium/tree/feature/lbipam-shared-ips

I also noticed that the history of this PR is getting a bit messy, we are not at 9 commits, not including these fixes. My suggestion is that you merge these commits into you own branch, then squash all code related commits into one commit (including my additions)  and let the docs commit be a separate commit.

usiegl00 committed:
lbipam: Add logic for requesting multiple ips with a sharing key

Updated the service compatibility check. Made the namespace check non
conditional since its a security check. Rewrote some of the logic to
be more consice. And added comments and rational to certain checks
to remind us of why we do them.

After the addition of the shared IP logic, some functions became to
large. This commit splits the code into smaller functions so the overall
flow is easier to follow.

When 2 services have a sharing key and the second service would update
so it is now incompatible with the other service it would keep the same
IP address due to a lack of compatibility checking during updates and
a logic bug in the ingress IP import code.

Add a test for sharing keys. Add logic for stacking multiple ips for one
sharing key. Add lbipam support for sharing keys.

Co-authored-by: Dylan Reimerink <dylan.reimerink@isovalent.com>
Signed-off-by: usiegl00 <50933431+usiegl00@users.noreply.github.com>

usiegl00 committed:
lbipam: Add documentation for sharing keys and shared ips.

Signed-off-by: usiegl00 <50933431+usiegl00@users.noreply.github.com>

dylandreimerink reviewed (approved):
The LB-IPAM logic looks good to me! :tada:

usiegl00 commented on Documentation/network/lb-ipam.rst:
```diff
@@ -429,3 +429,49 @@ for the network and broadcast addresses respectively.
+Sharing Keys
+------------
+
+Services can share the same ip or set of ips with other services. This is done by setting the ``io.cilium/lb-ipam-sharing-key`` annotation on the service.
```
Thanks! I've updated the pr.
